### PR TITLE
Draft - Introduce unit testing using google tests

### PR DESCRIPTION
**Pull request by royfalk**
This draft PR introduces unit tests based on gtests by google. 
If approved, I will begin introducing unit tests to the other ongoing PR's, primarily #484.

Code Changes:
- [ No ] Have the PR Validation Tests been run? See https://github.com/vegastrike/Vega-Strike-Engine-Source/wiki/Pull-Request-Validation
- [ ] This is a documentation change only

Issues:
- None

Purpose:
- What is this pull request trying to do?
- What release is this for? 0.9
- Is there a project or milestone we should apply this to?

### CODE REVIEW

File: engine/src/tests/test.cpp [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#include <gtest/gtest.h>
+
+// Demonstrate some basic assertions.
+TEST(HelloTest, BasicAssertions) {
+  // Expect two strings not to be equal.
+  EXPECT_STRNE("hello", "world");
+  // Expect equality.
+  EXPECT_EQ(7 * 6, 42);
+}
```
Comment (BenjamenMeyer) [inline engine/src/tests/test.cpp:9]: For comparison Boost::Test equivalent:

```C++
#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_NO_MAIN

#include <boost/test/unit_test.hpp>

BOOST_AUTO_TEST_SUITE( GroupName );

BOOST_AUTO_TEST( IndividualTestName )
{
    // For whatever reason Boost doesn't do a not-equal directly
    BOOST_CHECK("hello" != "world");
    // but they do do equal directly
    BOOST_CHECK_EQUAL(7*6, 42);
}
```

**Review by BenjamenMeyer — COMMENTED** [review]
Added an example of how it would be done using Boost::Test for comparison.
Haven't used Google's gtest before, so it's interesting to compare.

gtest looks a little simpler from some perspectives; though it'd be nice to see how it works with multiple C++ files and if there's anything special required for that.

**Comment by ministerofinformation** [comment]
Chatting with some folks who've worked on Google product teams, they're pretty happy with the Google testing infrastructure. I got the impression that the feature set is deep, especially in  the "mock" portions, but that the learning curve is appropriately proportional to the complexity of the codebase and behaviors being tested (specifically, if your code is more template metaprogramming than actual C++, then you'll have more fun understanding the associated errors coming out of the framework, but you also brought that complexity on yourself...)

**Comment by ministerofinformation** [comment]
whoops - sorry about that, mis-clicked while typing on my phone.

**Comment by Loki1950** [comment]
So it really depends on how much work we put into our testing paradigm and it's integration with our code base.

**Review by stephengtuggy — COMMENTED** [review]
I like this. gtest looks like a good option for unit testing.

BTW, the existing code does seem to have some unit tests of sorts, here and there. Nothing particularly organized though. It would be good to translate those into the unit testing framework that we end up settling on.

File: engine/src/tests/test.cpp (hunk 1 above)
Comment (royfalk) [reply engine/src/tests/test.cpp:9]: Yeah. Like I said, at the basic level, they're all the same and I have looked at Boost.

**Comment by royfalk** [comment]
@BenjamenMeyer I suggest we close this PR. 
#484 has cherry picked this commit and implemented some meaningful first unit tests.

**Comment by stephengtuggy** [comment]
Closing, per @royfalk 's request.

**Pull request closed without merging**